main/src/ca/uwaterloo/flix/language/phase/njvm/Mnemonics.scala

```diff
@@ -359,940 +638,1496 @@ object Mnemonics {
       t => t.emitNew(jt)
 
     /**
       * Duplicates whatever is on Top of the stack
       */
-    def DUP[R <: Stack, T: TypeTag]: F[R ** T] => F[R ** T ** T] =
+    def DUP[S <: Stack, T: TypeTag]: F[S ** T] => F[S ** T ** T] =
       t => t.emitDup()
 
     /**
       * Loads a constant string onto the stack
       */
-    def LDC_STRING[R <: Stack](value: String): F[R] => F[R ** JString.type] =
+    def LDC_STRING[S <: Stack](value: String): F[S] => F[S ** Ref[MString]] =
       t => t.emitLdc(value)
 
+    /**
+      * Loads a constant int onto the stack
+      */
+    def LDC_INT[S <: Stack](value: Int): F[S] => F[S ** MInt] =
+      t => t.emitLdc(value.asInstanceOf[Object])
+
+    /**
+      * Loads a constant int onto the stack
+      */
+    def CONST_NULL[S <: Stack, T <: Ref[MObject]]: F[S] => F[S ** T] =
+      t => t.emitConstNull
+
+    /**
+      * Loads a constant int onto the stack
+      */
+    def NEWARRAY[S <: Stack, T <: MObject : TypeTag](implicit root: Root, flix: Flix): F[S ** MInt] => F[S ** MArray[Ref[T]]] =
+      t => t.emitNewArray(getJvmType[Ref[T]].asInstanceOf[Reference])
+
     /**
       * Throws the exception which is currently on top of the stack
       */
-    def THROW: F[StackNil ** Reference] => F[StackNil] =
+    def THROW: F[StackNil ** Ref[MObject]] => F[StackNil] =
       t => t.emitThrow()
 
-    def IFEQ[S <: Stack](f: F[S] => F[S]): F[S ** PrimBool] => F[S] = {
+    def IFEQ[S <: Stack](f: F[S] => F[S]): F[S ** MBool] => F[S] = {
       val skipLabel = new Label
-      ((t: F[S ** PrimBool]) => t.emitIfeq[S](skipLabel)) |>>
+      ((t: F[S ** MBool]) => t.emitIfeq[S](skipLabel)) |>>
         f |>>
         (t => t.emitLabel(skipLabel))
     }
 
-    def CHECK_CAST[S <: Stack](jt: Reference): F[S ** Reference] => F[S ** Reference] =
+    def CHECK_CAST[S <: Stack](jt: Reference): F[S ** Ref[MObject]] => F[S ** Ref[MObject]] =
       t => t.emitCheckCast(jt)
 
     def POP[S <: Stack, T](): F[S ** T] => F[S] = ???
 
-  }
-
-  /**
-    * Capability which allows to load/store a local variable
-    */
-  class Local[T: TypeTag](location: Int)(implicit root: Root, flix: Flix) {
+    def NO_OP[S <: Stack]: F[S] => F[S] = t => t
 
-    private val localType = getJvmType[T]
-
-    def LOAD[R <: Stack]: F[R] => F[R ** T] = t => t.emitLoad(localType, location)
-
-    def STORE[R <: Stack]: F[R ** T] => F[R] = t => t.emitLoad(localType, location)
+    def AASTORE[S <: Stack, T <: Ref[MObject], T1 <: Ref[_]]: F[S ** MArray[T] ** MInt ** T1] => F[S] = t => t.emitAAStore
   }
 
-  /**
-    * Capability which allows to get/put a field
-    */
-  class Field[T: TypeTag](fieldName: String)(implicit root: Root, flix: Flix) {
-
-    private val fieldType = getJvmType[T]
+    /**
+      * Capability which allows to load/store a local variable
+      */
+    class Local[T <: MnemonicsTypes : TypeTag](location: Integer)(implicit root: Root, flix: Flix) {
```

Comment:
Location is a poor choice of name- we use that for source location. How about offset? or stackOffset? or variableNumber? etc?